### PR TITLE
tool/arxivsearch: limit PDF content results

### PR DESCRIPTION
**Pull request by WineChord**
Summary
- cap PDF text returned by arxiv_search when read_arxiv_papers is enabled
- add truncation metadata and an optional max_content_runes override
- keep normal metadata-only arXiv search behavior unchanged
- validate submitted-date helper fields and document open-ended date ranges

Behavior notes
- read_arxiv_papers=true now returns PDF text up to 20,000 runes per article by default, with truncation metadata when content is capped.
- max_content_runes can raise the per-article cap up to 200,000 runes; non-positive values use the default. Use read_arxiv_papers=false for metadata-only search results.
- submitted_date_from/submitted_date_to may be used for open-ended ranges, but inverted ranges and duplicate query-level submittedDate clauses are rejected.

Validation
- cd tool/arxivsearch && go test ./...

### CODE REVIEW

File: tool/arxivsearch/arxiv_search.go [hunk 1]
```diff
@@ -171,8 +185,11 @@ func NewToolSet(opts ...Option) (*ToolSet, error) {
 	return t, nil
 }
 
-func (t *ToolSet) search(ctx context.Context, req searchRequest) ([]article, error) {
-	if len(req.Search.Query) == 0 && len(req.Search.IDList) == 0 {
+func (t *ToolSet) search(
+	ctx context.Context,
+	req searchRequest,
+) ([]article, error) {
+	if isEmptySearch(req.Search) {
```
[the remark this answers is not in the document]
Comment (WineChord) [reply tool/arxivsearch/arxiv_search.go:196]: fixed

File: tool/arxivsearch/arxiv_search.go [hunk 2]
```diff
@@ -213,14 +233,80 @@ func (t *ToolSet) search(ctx context.Context, req searchRequest) ([]article, err
...
-				})
-			}
+			appendArticleContent(&arti, documents, maxContentRunes)
 		}
 		articles = append(articles, arti)
 	}
 	return articles, nil
 }
+
+func isEmptySearch(search arxiv.Search) bool {
```
Comment (liuzengh) [inline tool/arxivsearch/arxiv_search.go:245]: **Medium — bug**

`isEmptySearch` checks raw `len(...)`, but `buildSearchQuery` / `submittedDateClause` use `strings.TrimSpace`. Whitespace-only inputs slip through validation and produce an empty `search_query`:

```json
{"search": {"submitted_date_from": "   "}}
```

→ not rejected, but arXiv is called with no search criteria.

**Suggested fix:** trim before checking emptiness:

```go
func isEmptySearch(search arxiv.Search) bool {
	return strings.TrimSpace(search.Query) == "" &&
		len(search.IDList) == 0 &&
		strings.TrimSpace(search.SubmittedDateFrom) == "" &&
		strings.TrimSpace(search.SubmittedDateTo) == ""
}
```

Please add a test case for whitespace-only `query` / `submitted_date_from`.

File: tool/arxivsearch/arxiv_search.go [hunk 3]
```diff
@@ -34,32 +37,45 @@ var (
...
+	ContentTruncated     bool      `json:"content_truncated,omitempty"`
 }
 
 // searchRequest define an arxiv search request
 type searchRequest struct {
 	Search          arxiv.Search `json:"search" jsonschema:"description=Search query"`
 	ReadArxivPapers bool         `json:"read_arxiv_papers" jsonschema:"description=Whether to read the content from PDF"`
+	// MaxContentRunes limits PDF text returned per article.
+	// It defaults to 20000 and is capped at 200000.
+	MaxContentRunes int `json:"max_content_runes,omitempty"`
```
Comment (liuzengh) [inline tool/arxivsearch/arxiv_search.go]: **Medium — discoverability**

`max_content_runes` has a Go comment but no `jsonschema` tag. LLM tool callers won't see the override, default (20k), or cap (200k).

**Suggested fix:**

```go
MaxContentRunes int `json:"max_content_runes,omitempty" jsonschema:"description=Max PDF text runes returned per article when read_arxiv_papers is true. Defaults to 20000, capped at 200000."`
```

File: tool/arxivsearch/arxiv_search.go [hunk 4]
```diff
@@ -34,32 +37,45 @@ var (
 	maxResults = 5
 )
 
+const (
+	defaultArticleContentRunes = 20000
```
Comment (liuzengh) [inline tool/arxivsearch/arxiv_search.go:42]: **Low — note for PR description**

Worth calling out in the PR body: enabling `read_arxiv_papers` is a **behavior change** — responses that used to include full PDF text are now capped unless `max_content_runes` is raised. Metadata fields (`content_truncated`, `content_runes`, `returned_content_runes`) mitigate this, but downstream consumers should be aware.

File: tool/arxivsearch/internal/arxiv/client.go [hunk 5]
```diff
@@ -161,6 +165,86 @@ func (c *Client) buildQueryURL(search Search, start, maxResults int) (string, er
...
+	if err != nil {
+		return "", err
+	}
+	if dateClause == "" {
+		return query, nil
+	}
+	if query == "" {
+		return dateClause, nil
+	}
+	return query + " AND " + dateClause, nil
```
Comment (liuzengh) [inline tool/arxivsearch/internal/arxiv/client.go]: **Medium — footgun**

If `query` already contains `submittedDate:[...]` and the caller also sets `submitted_date_from` / `submitted_date_to`, both clauses are ANDed:

```
cat:cs.AI AND submittedDate:[20220101 TO 20221231] AND submittedDate:[202206010000 TO 202206302359]
```

That often returns zero results with no error.

**Suggested fix (pick one):**
- Document in jsonschema: *"Do not embed submittedDate in query when using submitted_date_from/to."*
- Detect duplicate `submittedDate` in `query` when date fields are set and return an error.
- Prefer structured fields and strip/ignore date clauses in `query` (more invasive).

File: tool/arxivsearch/internal/arxiv/client.go [hunk 6]
```diff
@@ -161,6 +165,86 @@ func (c *Client) buildQueryURL(search Search, start, maxResults int) (string, er
...
+	}
+	fromDate, err := normalizeArxivDateBound(from, false)
+	if err != nil {
+		return "", fmt.Errorf("invalid submitted_date_from: %w", err)
+	}
+	toDate, err := normalizeArxivDateBound(to, true)
+	if err != nil {
+		return "", fmt.Errorf("invalid submitted_date_to: %w", err)
+	}
+	return fmt.Sprintf(
```
Comment (liuzengh) [inline tool/arxivsearch/internal/arxiv/client.go:225]: **Low — UX**

There is no check that `fromDate <= toDate`. Inverted ranges like `from=2022-12-31`, `to=2022-01-01` build a valid query but return empty results silently.

**Suggested fix:**

```go
if fromDate > toDate {
	return "", fmt.Errorf("submitted_date_from must be before or equal to submitted_date_to")
}
```

Add a test in `client_test.go`.

File: tool/arxivsearch/internal/arxiv/types.go [hunk 7]
```diff
@@ -68,11 +68,13 @@ type Result struct {
...
-	SortOrder  SortOrder     `json:"sort_order,omitempty" jsonschema:"description=Sort order: ascending or descending,enum=ascending,enum=descending"`
+	Query             string        `json:"query,omitempty" jsonschema:"description=The search query string for arXiv articles"`
+	IDList            []string      `json:"id_list,omitempty" jsonschema:"description=List of arXiv IDs to search for"`
+	MaxResults        *int          `json:"max_results,omitempty" jsonschema:"description=Maximum number of results to return"`
+	SortBy            SortCriterion `json:"sort_by,omitempty" jsonschema:"description=Sort criterion: relevance or lastUpdatedDate or submittedDate,enum=relevance,enum=lastUpdatedDate,enum=submittedDate"`
+	SortOrder         SortOrder     `json:"sort_order,omitempty" jsonschema:"description=Sort order: ascending or descending,enum=ascending,enum=descending"`
+	SubmittedDateFrom string        `json:"submitted_date_from,omitempty" jsonschema:"description=Submitted date lower bound in YYYY-MM-DD or YYYYMMDDHHMM GMT"`
```
Comment (liuzengh) [inline tool/arxivsearch/internal/arxiv/types.go]: **Low — documentation**

Consider noting in jsonschema:
- Open-ended ranges are supported (only `from` → upper bound `999912312359`; only `to` → lower bound `199001010000`).
- Prefer these fields over embedding `submittedDate:[...]` in `query`.

Example:

```go
SubmittedDateFrom string `json:"submitted_date_from,omitempty" jsonschema:"description=Submitted date lower bound (YYYY-MM-DD or YYYYMMDDHHMM GMT). If omitted, defaults to 1990-01-01. Do not also embed submittedDate in query."`
```

File: tool/arxivsearch/arxiv_search.go [hunk 8]
```diff
@@ -213,14 +233,80 @@ func (t *ToolSet) search(ctx context.Context, req searchRequest) ([]article, err
...
+
+func isEmptySearch(search arxiv.Search) bool {
+	return len(search.Query) == 0 &&
+		len(search.IDList) == 0 &&
+		len(search.SubmittedDateFrom) == 0 &&
+		len(search.SubmittedDateTo) == 0
+}
+
+func normalizedMaxContentRunes(maxRunes int) int {
+	if maxRunes <= 0 {
```
Comment (liuzengh) [inline tool/arxivsearch/arxiv_search.go:262]: **Nit — design clarity**

`max_content_runes <= 0` always maps to the 20k default, so callers cannot intentionally request zero content. That seems fine for this tool; consider a one-line comment that `0` means "use default", not "return no content", to avoid confusion.

File: tool/arxivsearch/arxiv_search_test.go [hunk 9]
```diff
@@ -494,6 +523,46 @@ func TestWithTimeout_EnforcedAtCallPath(t *testing.T) {
 		"original transport must remain unmutated")
 }
 
+func TestNormalizedMaxContentRunes(t *testing.T) {
```
Comment (liuzengh) [inline tool/arxivsearch/arxiv_search_test.go:568]: **Low — missing tests**

Unit tests for `appendArticleContent` / `normalizedMaxContentRunes` look good. Gaps worth covering before merge:

| Scenario | Status |
|---|---|
| End-to-end truncation via `search()` + `read_arxiv_papers` + low `max_content_runes` | Missing |
| Pages fully skipped after budget exhausted (not only partial last page) | Missing |
| Open-ended date bounds (only `from` / only `to`) in `buildQueryURL` | Missing |
| Whitespace-only `query` / date fields rejected by `isEmptySearch` | Missing |

Codecov reports ~87% patch coverage; these would close the main gaps.

File: tool/arxivsearch/arxiv_search.go [hunk 10]
```diff
@@ -34,32 +37,45 @@ var (
...
+	Content              []content `json:"content"`
+	ContentRunes         int       `json:"content_runes,omitempty"`
+	ReturnedContentRunes int       `json:"returned_content_runes,omitempty"`
+	ContentTruncated     bool      `json:"content_truncated,omitempty"`
 }
 
 // searchRequest define an arxiv search request
 type searchRequest struct {
 	Search          arxiv.Search `json:"search" jsonschema:"description=Search query"`
 	ReadArxivPapers bool         `json:"read_arxiv_papers" jsonschema:"description=Whether to read the content from PDF"`
```
Comment (liuzengh) [inline tool/arxivsearch/arxiv_search.go:76]: **Medium — breaking change documentation**

When `read_arxiv_papers` is true, PDF text is now truncated at 20,000 runes by default. The tool description still says only "content (if read_arxiv_papers is true)" with no mention of truncation or `max_content_runes` / `content_truncated`.

Existing callers may assume full PDF text.

**Suggested fix:** extend `WithDescription`, e.g.:

```go
"and content (if read_arxiv_papers is true; PDF text is truncated per article, default 20000 runes, override via max_content_runes up to 200000; see content_truncated and content_runes metadata)."
```

**Review by liuzengh — COMMENTED** [review]
## Review summary

Solid truncation implementation and good Unicode coverage in unit tests. A few items to address before merge:

1. **Whitespace-only search criteria** bypass `isEmptySearch` (see inline on `isEmptySearch`)
2. **Document the breaking change** for `read_arxiv_papers` + add `jsonschema` for `max_content_runes`
3. **Duplicate `submittedDate`** when using both `query` and structured date fields
4. **Additional tests** for e2e truncation and open-ended date bounds

**Scope note:** consider updating the PR title/description to mention submitted-date filtering alongside PDF content limits.

Happy to re-review after updates.

File: tool/arxivsearch/arxiv_search.go (hunk 2 above)
Comment (WineChord) [reply tool/arxivsearch/arxiv_search.go:245]: fixed

File: tool/arxivsearch/arxiv_search.go (hunk 3 above)
Comment (WineChord) [reply tool/arxivsearch/arxiv_search.go]: fixed

File: tool/arxivsearch/arxiv_search.go (hunk 4 above)
Comment (WineChord) [reply tool/arxivsearch/arxiv_search.go:42]: fixed

File: tool/arxivsearch/internal/arxiv/client.go (hunk 5 above)
Comment (WineChord) [reply tool/arxivsearch/internal/arxiv/client.go]: fixed

File: tool/arxivsearch/internal/arxiv/client.go (hunk 6 above)
Comment (WineChord) [reply tool/arxivsearch/internal/arxiv/client.go:225]: fixed

File: tool/arxivsearch/internal/arxiv/types.go (hunk 7 above)
Comment (WineChord) [reply tool/arxivsearch/internal/arxiv/types.go]: fixed

File: tool/arxivsearch/arxiv_search.go (hunk 8 above)
Comment (WineChord) [reply tool/arxivsearch/arxiv_search.go:262]: fixed

File: tool/arxivsearch/arxiv_search_test.go (hunk 9 above)
Comment (WineChord) [reply tool/arxivsearch/arxiv_search_test.go:568]: fixed

File: tool/arxivsearch/arxiv_search.go (hunk 10 above)
Comment (WineChord) [reply tool/arxivsearch/arxiv_search.go:76]: fixed

File: tool/arxivsearch/internal/arxiv/client.go [hunk 11]
```diff
@@ -161,6 +165,105 @@ func (c *Client) buildQueryURL(search Search, start, maxResults int) (string, er
...
+	if containsSubmittedDateClause(query) {
+		return "", fmt.Errorf(
+			"query already contains submittedDate; use either the " +
+				"query clause or submitted_date_from/to",
+		)
+	}
+	if query == "" {
+		return dateClause, nil
+	}
+	return query + " AND " + dateClause, nil
```
Comment (Rememorio) [inline tool/arxivsearch/internal/arxiv/client.go]: **P1: Group the original query before applying the date filter**

Appending the clause directly changes the meaning of compound queries. For example, `cat:cs.AI OR cat:cs.LG` becomes `cat:cs.AI OR cat:cs.LG AND submittedDate:[...]`; because `AND` binds more tightly, the date range applies only to the second branch, so results outside the requested range can be returned.

Wrap a non-empty original query before adding the date clause, producing `(cat:cs.AI OR cat:cs.LG) AND submittedDate:[...]`, and add a regression assertion covering an `OR` query.

<details><summary>中文</summary>

**应用日期过滤前应对原查询分组**

直接追加日期子句会改变复合查询的语义。例如，`cat:cs.AI OR cat:cs.LG` 会变成 `cat:cs.AI OR cat:cs.LG AND submittedDate:[...]`。由于 `AND` 的优先级更高，日期范围只约束第二个分支，因此可能返回日期范围之外的论文。

建议先对非空原查询加括号，再拼接日期条件，生成 `(cat:cs.AI OR cat:cs.LG) AND submittedDate:[...]`，并增加一个包含 `OR` 查询的回归断言。

</details>

File: tool/arxivsearch/arxiv_search.go [hunk 12]
```diff
@@ -34,32 +38,45 @@ var (
...
+	// MaxContentRunes limits PDF text returned per article.
+	// It defaults to 20000 and is capped at 200000.
+	MaxContentRunes int `json:"max_content_runes,omitempty" jsonschema:"description=Maximum PDF text runes returned per article when read_arxiv_papers is true. Non-positive values use the 20000 default; values above 200000 are capped. Set read_arxiv_papers to false to omit PDF content."`
```
Comment (Rememorio) [inline tool/arxivsearch/arxiv_search.go:79]: **P2: Document the new partial-content response in the example**

Previously, `read_arxiv_papers=true` returned all extracted page text; it now returns at most 20,000 runes per article by default. The arXiv example README still presents `content` as complete page text and omits `max_content_runes` and the truncation metadata, so users can unknowingly treat partial paper text as complete.

Update the example input/output and feature description to show the default cap, the override limit, and how `content_truncated`, rune counts, and per-page `truncated` indicate partial results.

<details><summary>中文</summary>

**在示例中说明新的部分内容响应**

此前 `read_arxiv_papers=true` 会返回全部提取出的页面文本；现在默认每篇文章最多返回 20,000 个 rune。现有 arXiv 示例 README 仍将 `content` 描述为完整页面文本，也没有展示 `max_content_runes` 和截断元数据，用户可能误把不完整的论文内容当作全文。

请更新示例输入、输出和功能说明，明确默认上限、可覆盖的最大值，以及如何通过 `content_truncated`、rune 计数和页面级 `truncated` 判断结果是否被截断。

</details>

**Review by Rememorio — COMMENTED** [review]
I reviewed the changed lines and left 2 inline comments below. The comments focus on issues worth addressing before merge.

Reviewed PDF content budgeting, truncation metadata, submitted-date helpers, query composition, and related tests. The content cap is implemented consistently, but compound queries can escape the new date filter, and the repository example should document the intentional default behavior change.

<details><summary>中文</summary>

我看过这次变更的相关代码，在下面留下 2 条行内评论。评论聚焦在合并前值得处理的问题。

已审查 PDF 内容预算、截断元数据、提交日期辅助字段、查询组合及相关测试。内容上限的实现基本一致，但复合查询可能绕过新增日期过滤，并且仓库示例需要说明这一有意的默认行为变更。

</details>

File: tool/arxivsearch/internal/arxiv/client.go (hunk 11 above)
Comment (WineChord) [reply tool/arxivsearch/internal/arxiv/client.go]: fixed

File: tool/arxivsearch/arxiv_search.go (hunk 12 above)
Comment (WineChord) [reply tool/arxivsearch/arxiv_search.go:79]: fixed

File: tool/arxivsearch/internal/arxiv/client.go (hunk 11 above)
Comment (Rememorio) [reply tool/arxivsearch/internal/arxiv/client.go]: Confirmed fixed: the original query is now grouped before the date filter is appended, and the OR-query regression test passes.

<details><summary>中文</summary>

已确认修复：原查询现在会先分组再拼接日期过滤条件，OR 查询回归测试已通过。

</details>

File: tool/arxivsearch/internal/arxiv/client.go [hunk 13]
```diff
@@ -161,6 +165,105 @@ func (c *Client) buildQueryURL(search Search, start, maxResults int) (string, er
...
+		toDate,
+	), nil
+}
+
+func normalizeArxivDateBound(raw string, upper bool) (string, error) {
+	raw = strings.TrimSpace(raw)
+	if raw == "" {
+		return "", fmt.Errorf("empty date")
+	}
+	digits := dateDigits(raw)
```
Comment (Rememorio) [inline tool/arxivsearch/internal/arxiv/client.go]: **P2: Reject malformed dates instead of stripping arbitrary characters**

`dateDigits` removes every non-digit before validation, so malformed values such as `2022-0x6-01` are silently accepted as `2022-06-01`. This contradicts the documented formats and can return results for a date different from what the caller supplied.

Parse the explicitly supported layouts directly, optionally adding a specific layout for the formatted timestamp covered by the tests. Please also add a negative case containing an unexpected character within a date component.

<details><summary>中文</summary>

**应拒绝格式错误的日期，而不是任意剥离字符**

这里先通过 `dateDigits` 删除所有非数字字符，再进行日期校验，因此像 `2022-0x6-01` 这样的错误输入会被静默当作 `2022-06-01`。这不符合文档声明的格式，也可能让调用方得到与原始输入不一致的日期范围结果。

建议直接按明确支持的日期格式解析；如果需要支持测试中的带分隔符时间戳，也应为它定义单独的格式。同时补充一个日期字段内部含意外字符的反向测试。

</details>

**Review by Rememorio — COMMENTED** [review]
I reviewed the changed lines and left 1 inline comment below. The comments focus on issues worth addressing before merge.

The PDF rune cap preserves metadata-only behavior and includes focused truncation tests. One date-validation issue remains: malformed input can be silently normalized into a different valid date.

<details><summary>中文</summary>

我看过这次变更的相关代码，在下面留下 1 条行内评论。评论聚焦在合并前值得处理的问题。

PDF 字符上限保持了纯元数据查询的原有行为，并补充了针对截断逻辑的测试。目前仍有一个日期校验问题：格式错误的输入可能被静默转换为另一个有效日期。

</details>

File: tool/arxivsearch/internal/arxiv/client.go (hunk 13 above)
Comment (WineChord) [reply tool/arxivsearch/internal/arxiv/client.go]: fixed in 35883b508: supported layouts are now parsed explicitly, malformed embedded characters are rejected, and a 2022-0x6-01 regression case was added

**Review by Flash-LHR — APPROVED** [review]
LGTM
